<a href="https://colab.research.google.com/github/cedeerwe/brutalna-akademia/blob/master/notebooks/06_minca_cloveku.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pomocný kód

Spustite všetko tlačidlom *Runtime -> Run all*.

Poprosíme neotvárať a nepozerať až po začatie domácej úlohy.



In [0]:
import numpy as np
import pandas as pd
from io import StringIO
import plotly.graph_objects as go

np.random.seed(0)

def get_data():
    data = pd.read_csv("https://raw.githubusercontent.com/cedeerwe/brutalna-akademia/master/data/minca_cloveku1.csv", index_col=0)
    data.columns = data.columns.map(int)
    return data

def f1_score(positives):
    if (len(positives) == 0):
        print("At least one guess is necessary!")
        return 0
    solutions = {0, 3, 4, 5, 8, 13, 17, 21, 23, 26, 29, 32, 39, 41, 43, 46, 48, 49, 50, 54, 55, 57, 
                 58, 61, 65, 67, 68, 70, 71, 75, 77, 78, 79, 80, 82, 84, 87, 89, 97, 98, 100, 101, 
                 103, 106, 109, 110, 111, 113, 115, 117, 120, 122, 123, 124, 127, 129, 130, 135, 
                 138, 139, 140, 141, 143, 144, 150, 153, 162, 164, 165, 168, 169, 170, 171, 173, 
                 177, 178, 179, 181, 182, 185, 188, 189, 190, 193, 195, 199, 200, 202, 203, 205, 
                 211, 214, 216, 220, 225, 232, 236, 240, 241, 244, 248, 249, 250, 252, 254, 255, 
                 259, 260, 262, 267, 269, 271, 272, 273, 275, 276, 280, 285, 290, 291, 292, 293, 
                 295, 298, 299, 302, 303, 307, 314, 322, 328, 331, 334, 336, 340, 341, 343, 344, 
                 345, 347, 348, 350, 353, 357, 358, 362, 364, 365, 366, 369, 372, 373, 374, 375, 
                 377, 380, 382, 383, 387, 392, 394, 396, 399, 402, 404, 406, 414, 417, 418, 420, 
                 421, 423, 424, 429, 430, 435, 436, 438, 439, 442, 449, 450, 452, 456, 458, 463, 
                 465, 468, 471, 472, 473, 477, 481, 483, 484, 488, 491, 496, 497, 498}
    set_positives = set([int(i) for i in positives])
    correct = solutions.intersection(set_positives)
    precision = len(correct) / len(set_positives)
    recall = len(correct) / len(solutions)
    if precision == 0 and recall == 0:
        return 0
    return 2 * precision * recall / (precision + recall)

def get_data_homework():
    return pd.read_csv("https://raw.githubusercontent.com/cedeerwe/brutalna-akademia/master/data/minca_cloveku2.csv", header=None).T

def get_data_practical():
    data = pd.read_csv("https://raw.githubusercontent.com/cedeerwe/brutalna-akademia/master/data/thyroid.csv", header=None)
    return pd.DataFrame(data.iloc[:, :-1])

# Úloha č.1

Dnešná prvá úloha bude interaktívna, odporúčame ju vykonať v spoločnosti aspoň niekoľkých ľudí. Nasleduje inštrukcia pre každého z vás, vykonávajte ich individuálne.
1. Pripravte si papier, pero a mincu.
2. Nájdite si miesto, kde vám nikto nebude pozerať do papiera. 




## Pokračovanie


3. *Predstavte si*, že ste hodili mincou a zapíšte výsledok takéhoto hodu na papier ako písmeno H (hlava) alebo Z (znak).

## Pokračovanie

4. Zopakujte krok 3., aby ste mali dokopy napísaných 30 písmen vedľa seba. Očakávame teda niečo ako HZHHZHZZZ... Snažte sa, aby váš reťazec 30 písmen vyzeral čo najviac ako výsledky hodov mincou, bez použitia mince samotnej. 


## Pokračovanie

5. Zoberte mincu a hoďte ju 30 krát. Zapisujte si priebežne výsledky.


## Pokračovanie

6. Prepíšte si niekde tieto dva reťazce 30 hodov mince (skutočnej a vymyslenej) vedľa seba tak, aby nebolo jasné, ktorá je pravá a ktorá je vymyslená.
7. Navštívte postupne ľudí vo vašom okolí a dajte im úlohu, aby určili, ktorá z týchto dvoch postupností je vymyslená a ktorá predstavuje skutočné výsledky hodou mincou. Zapisujte si ich odpovede.

## Pokračovanie

8. Vyhodnoťte si ako to išlo a podľa čoho ste určovali, ktorá minca je pravá a ktorá je vymyslená.

## Zamyslenie na záver


Aj keď vyzerá byť na prvý pohľad úloha len ako jednoduchá hra, má veľmi blízko k aplikáciám v oblasti **detekcie anonomálií**. Táto oblasť má veľa zaujímavých použití, od detekcie zneužitia vášho online účtu, bankovej karty, cez filtrovanie spamových emailov, až po detekciu násilného správania z kamerových záznamov.

Naša predchádzajúca úloha a všetky tieto aplikácie majú spoločný jeden aspekt: snažíme sa rozlíšiť bežnú udalosť od udalosti niečím špeciálnej. V našej úlohe sme rozlišovali bežnú mincu od mince, ktorá nie je bežná, lebo ju vymýšľal človek. Podobne chceme rozlíšiť bežnú tranzakciu od krádeže alebo bežný pohyb občanov po meste od fyzického napadnutia. V žiadnej z týchto nie úloh nie je jednoduché povedať, čo presne definuje túto "nebežnú" udalosť. Vo všetkých z týchto príkladov vieme ale jasne povedať, že je táto udalosť dostatočne odlišná od udalosti bežnej. A práve na to sa sústredí detekcia anomálií a práve na to sa budeme dneska sústrediť. Budeme sa snažiť nájsť situácie, ktoré sú dostatočne odlišné od tých bežných. Ako to presne robiť, na to už budete musieť prísť vy.

# Úloha č.2

Podobne ako v prvej úlohe, budeme sa snažiť rozlíšiť medzi skutočne náhodnou sekvenciou hodov mincí a medzi ľudmi vytvorenými sekvenciami.


## Podrobnosti

In [2]:
get_data()

0    1    2    3    4    5    6    ...  493  494  495  496  497  498  499
0     1    0    0    1    1    1    1  ...    1    0    0    0    0    0    1
1     1    0    1    0    0    1    0  ...    1    0    0    1    0    0    1
2     1    1    1    0    0    1    1  ...    0    0    1    1    0    1    0
3     1    1    1    0    1    0    0  ...    1    1    0    0    0    1    1
4     0    0    1    1    0    0    0  ...    0    1    0    1    1    0    0
..  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...
95    1    0    0    0    1    1    1  ...    0    1    1    1    0    0    0
96    1    0    1    1    1    0    1  ...    1    1    0    0    0    0    0
97    0    1    1    1    1    0    1  ...    1    1    0    1    0    0    1
98    0    1    1    1    1    1    1  ...    1    1    0    0    0    1    0
99    1    1    1    0    0    0    0  ...    1    0    0    0    1    0    0

[100 rows x 500 columns]

Príkazom `get_data()` sa viete dostať k zozbieraným dátam k tejto problematike. Každý stĺpec je jeden respondent a každé políčko reprezentuje jeden jeho vymyslený / náhodný hod mincou. Dokopy máme 500 respondentov o dĺžke sekvencií 100. Vašou úlohou je identifikovať, ktorý z týchto respondentov je človek.

Respondenti sú označení číslami `0-499`. Akonáhle budete mať tip na čísla ľudských respondentov, môžte si overiť svoj tip pomocou funkcie `f1_score`. Táto funkcia berie ako vstup zoznam čísel a vracia vašu presnosť ako číslo medzi 0 (absolútne nepresné) a 1 (absolútne presné). Napríklad:




In [3]:
f1_score([10, 20, 30, 40, 50, 60, 70, 80, 90, 100])

0.03809523809523809

Nemáme nulovú presnosť, takže pravdepodobne aspoň jedno z týchto čísel bude správne. Tento výsledok je počítaný metrikou známou ako **F1 score**, o ktorej si môžte prečítať viac [tu](https://en.wikipedia.org/wiki/F1_score).

Ak by ste riešili reálny problém, takúto funkciu na overovanie vášho výsledku by ste nemali k dispozícií. Prosíme vás preto, aby ste nezneužívali dostupnosť tejto funkcie na spresnenie vášho výsledku. Napríklad je veľmi ľahké napísať program, ktorý postupne pridáva čísla do tejto postupnosti a necháva si iba tie, ktoré zvýšia presnosť.

Vo všeobecnosti vaša snaha o čo najlepšiu presnosť bude vyzerať zhruba takto:
1. Navrhnem si spôsob, akým určím čísla ľudských respondentov.
2. Použijem tento spôsob na získanie výsledku.
3. Pozriem si akú dosahujem presnosť.
4. Opakujem kroky 1.-3.

Ak vám nie je jasné, čo už znamená zneužívanie dostupnosti funkcie `f1_score`, môžte sa riadiť nasledovným pravidlom: Jedno takéto kolečko by vám malo trvať *minúty*, nie *sekundy*.

## Pomoc s prácou s dátami

Dáta sú načítané ako štruktúra, ktorá sa nazýva `DataFrame` z knižnice `pandas`. My máme načítanú celú knižnica `pandas` a uložili sme ju pod jej tradičnú skratku `pd`. Ak by ste chceli použiť nejakú funkciu z knižnice `pandas`, napríklad si vytvoriť vlastný `DataFrame`, môžte tak urobiť pomocou `pd.DataFrame()`.

K stĺpcom sa viete dostať jednoduchým indexovaním. Každý stĺpec je štruktúra, ktorá sa nazýva `Series`. 

Ak budete chcieť spraviť nejakú operáciu nad týmito štruktúrami, stačí použiť google a určite nájdete mnoho riešení. Napríklad ak by som chcel spočítať počet jednotiek v stĺpci, vygooglim `pandas series count` a dostanem sa k príkazu `.value_counts()`.

In [0]:
data = get_data() # uložíme si dáta do premennej, aby sme ich nemuseli vždy načítať

In [5]:
data[1] # jednoduché indexovanie číslom nám vráti stĺpec

0     0
1     0
2     1
3     1
4     0
     ..
95    0
96    0
97    1
98    1
99    1
Name: 1, Length: 100, dtype: int64

In [6]:
data[1].value_counts() # dostaneme počet jednotiek aj núl

1    59
0    41
Name: 1, dtype: int64

In [7]:
data[1].value_counts()[1] # počet jednotiek získame pod indexom 1

59

In [8]:
list(data[1]) # zobrazíme si všetky elementy

[0,
 0,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 1,
 1]

## Riešenie

K riešeniu takejto úlohy sa dá pristupovať viacerými spôsobmi. My si načrtneme jeden celý typ riešení, vyskúšame si ho použiť a necháme vám priestor, aby ste si ho sami dotiahli tak ďaleko, ako budete chcieť. Riešenie bude rozdelené na viacero častí a pred každým z nich budete mať priestor na to, aby ste sa znova zamysleli a skúsili si ďalši krok spraviť sami. Omnoho viac sa naučíte o problematike, keď sa ňou zaoberáte sami, ako keď si len prečítate riešenia.

### Štatistika

Na úvod si povieme niečo o pojme **štatistika**. Určite ste veľakrát toto slovo počuli, ale čo vlastne presne znamená? Na jednej strane (od anglického *statistics*) ide o obor, ktorý sa zaoberá skúmaním dát a ich vzťahov. Existuje ale aj druhý význam, ktorý je prekladom anglického *statistic*, a práve týmto pojmom sa budeme dnes zaoberať. Ak sa spýtame anglickej wikipedie, dostaneme sa k nasledovnej definícii:

*A statistic (singular) or sample statistic is any quantity computed from values in a sample, often the mean. Technically speaking, a statistic can be calculated by applying any mathematical function to the values found in a sample of data.*

Ľudovo povedané, štatistika je ľubovoľné číslo spočítane na základe našich dát. Predstavte si, že máme k dispozícií dáta o váhe nejakej skupiny ľudí, konkrétne 

$$
[45, 73, 115, 83, 109, 55, 63, 90, 45, 67].
$$

Príklady štatistík na týchto dátach môžu byť:
- počet meraní: 10,
- najväčšie meranie: 115,
- najmenšie meranie: 45,
- druhé najväčšie meranie: 109,
- počet meraní menších ako 70: 5,
- počet meraní rovných 45: 2,
- počet meraní rovných 46: 0,
- priemer: 74.5,
- geometrický priemer: 70.9,
- medián: 60,
- počet meraní deliteľných piatimi: 5,
- počet prvočísel: 3,
- najväčší nárast dvoch po sebe idúcich meraní: 42
- ...

Mohli by sme pokračovať naozaj dlho. Štatistikou môže byť čokoľvek, ale iba niektoré štatistiky sú aj užitočné. Napríklad počet prvočísel v meraniach váhy je pravdepodobne úplne zbytočná štatistika - nevieme na základe nej povedať o našich dátach nič užitočné. Na druhú stranu, priemer je väčšinou užitočnou štatistikou.

Keď sa pozrieme na naše dáta, môžme si na nich zobraziť rôzne štatistiky. Napríklad počet jednotiek pre každého z našich respondentov. 

In [9]:
data.apply(lambda col: (col == 1).sum())

0      52
1      59
2      56
3      53
4      52
       ..
495    56
496    46
497    47
498    47
499    53
Length: 500, dtype: int64

In [10]:
fig = go.Figure(data=[go.Histogram(x=data.apply(lambda col: (col == 1).sum()))])
fig.update_layout(
    title="Rozdelenie počtu vrchov v rámci respondentov",
    xaxis_title="Počet vrchov",
    yaxis_title="Počet respondentov",
)
fig.show()

Okrem toho, že si vieme počet jednotiek vo výpovediach našich respondentov ukázať, môžme si aj presne spočítať, aký by mal byť počet jednotiek pre skutočné mince.

Podúloha: Aká je šanca, že počet hláv z $n$ hodov mincou bude presne $k$?



#### Riešenie podúlohy

Pravdepodobnosť spočítame, ako počet vyhovujúcich možností vydelených počtom všetkých možností.

Počet všetkých možností, ako môže dopadnúť $n$ hodov mincou, bude $2^n$. Každá minca môže dopadnúť dvoma rôznymi spôsobmi a hádžeme dokopy $n$-krát.

Počet vyhovujúcich možností je počet možností, keď hodíme $k$-krát hlavu z $n$-hodov. Počet týchto možností je rovnaký, ako odpoveď na otázku, koľko spôsobmi môžme vybrať $k$-guličiek z $n$-guličiek, čoho odpoveďou je ${n} \choose{k}$.

Odpoveďou na podúlohu je teda 

$$
\frac{n \choose k}{2^n}
$$

#### Pokračovanie

Keď máme spočítané pravdepodobnosti na počty hláv, môžme si ich vyobraziť do grafu.

In [11]:
import math

N = 100
N_RESPONDENTS = 500
# histogram spravený rovnako ako v predchádzajúcej časti
trace_histogram = go.Histogram(
    x=data.apply(lambda col: (col == 1).sum()),
    histnorm='probability',
    name='histogram') 
# spočítame vzorec pre každú z hodnôt
x_true_probs = list(range(N + 1))
y_true_probs = [math.factorial(N) / math.factorial(N-k) / math.factorial(k) / (2 ** N) 
                for k in x_true_probs]
# vykreslíme ako čiaru
trace_true_probs = go.Scatter(
    x=x_true_probs,
    y=y_true_probs,
    mode='lines',
    name='true probability'
)
fig = go.Figure(data=[trace_histogram, trace_true_probs])
fig.update_layout(
    title="Rozdelenie počtu vrchov v rámci respondentov",
    xaxis_title="Počet vrchov",
    yaxis_title="Pravdepodobnosť / zmeraná pravdepodobnosť",
)
fig.show()

Na tomto príklade môžme napríklad vidieť, že v našich dátach sa častejšie vyskytuje 50 hláv, ako by sa vyskytovať malo. Dáva to zmysel - mnohí respondenti si pravdepodobne dávali pozor, aby dali rovnaký počet hláv a spodov, aby ich sekvencia vyzerala ako minca. V skutočnosti je počet hláv presne 50 iba v 8% prípadoch, čo je celkom málo.

Ukázali sme si, ako sa dá využiť štatistika, aby sme zistili niečo o našich dátach. Máte teraz priestor skúsiť využiť tieto poznatky, aby ste sa posunuli ďalej. Keď budete pripravený, pustíme sa do ďalšej časti.

### Pokročilejšia štatistika

Aj keď nám naša prvá štatistika dala nejakú intuíciu, nepomohla nám priamo identifikovať ľudských respondentov. Na našej vzorke si ľudia očividne dávali pozor, aby mali počet hláv v poriadku, ale zvládli si ustrážiť aj rozumný počet komplikovanejších štatistík? Uvidíme.

V tejto časti sa pozrieme na nasledovnú štatistiku: počet dvoch hláv idúcich po sebe. V prvom rade si skúsme spočítať, aký počet dvoch po sebe idúcich hláv by sme očakávali od skutočnej mince. Konkrétne, aká je šanca, že z $n$ hodov padne $k$ po sebe idúcich dvoch hláv? Pre upresnenie, sekvencia `HHHH` v sebe obsahuje 3 po sebe idúce dve hlavy.

#### Riešenie podúlohy

Ak ste sa zamysleli nad týmto problémom, pravdepodobne ste dospeli k záveru, že to vôbec nie je také jednoduché, ako sa na prvý pohľad zdá. To je správne pozorovanie. Vyriešenie tohto príkladu čisto s použitím matematiky je celkom náročné. 

Kľúčom k riešeniu tejto úlohy je matematická indukcia. Predstavte si, že máte riešenie pre $n$. Viete na základe neho nájsť riešenie pre $n+1$?

#### Riešenie podúlohy II.

Úloha je aj s použitím matematickej indukcie stále celkom náročná, ale už sa dá doriešiť. Aby sme vedeli rozumne spraviť krok z $n$ na $n+1$, potrebujeme si zapamätať výsledky pre všetky $k$ (počty dvoch hláv idúcich po sebe) a navyše aj ich rozdelenie podľa toho, *či bol posledný hod hlava alebo znak*. Prečo? Skúsme si spraviť prvých pár krokov a rýchlo to uvidíme.

* `n=1`, máme dve postupnosti s 0 dvojicami hlava-hlava. Jedna z nich končí na znak a jedna z nich končí na hlavu. Tento stav si zapíšeme ako `Z0 = 1, H0 = 1` - jedna postupnosť s 0 dvojicami `HH` končiaca na znak a jedna postupnosť s 0 dvojicami `HH` končiaca na hlavu.
* `n=2`, zoberieme si stav z `n=1`. Pozrieme sa na postupnosti v košiku `Z0`:
  * Každá z týchto postupností si môže na svoj koniec pridať buď hlavu alebo znak. Pozrieme sa na to individuálne:
    * Ak si `Z0` pridá na koniec hlavu, bude z nej `H0`. Nepridala sa jej žiadna dvojica `HH`, ale už má na konci hlavu. Z postupnosti tvaru `...xxxZ` sa stala `...xxxZH`, čo sa nám zaradí do kategórie `...xxxH`. 
    * Po pridaní znaku na koniec `Z0` dostaneme zase iba `Z0`.
  * Rovnako si rozoberieme kategóriu `H0`:
    * Ak si k nej pridáme hlavu, dostaneme sa do `H1`. Postupnosť `...xxxH` sme si rozšírili na `...xxxHH`, čím sa nám vytvorila jedna nová dvojica `HH` a stále máme na konci hlavu. 
    * Pridanie znaku na koniec `H0` vytvorí `Z0`
  * Ak si to celé zhrnieme, tak na dosiahnutie `n=2` spravíme operácie 
  ```
  Z0 <- Z0 + H0,
  H0 <- Z0,
  H1 <- H0.
  ```
  Ľavé strany týchto operácií zodpovedajú novému počtu postupností v košíkoch, kým pravé strany týchto operácií nám dávajú návod, ako ľavé strany spočítať. Použitím tohto návodu dostaneme stav `Z0 = 2, H0 = 1, H1 = 1`. Tento stav zodpovedá pravdepodobnosti $\tfrac{3}{4}$, že bude 0 dvojíc `HH` a pravdepodobnosti `\tfrac{1}{4},` že bude 1 dvojica `HH`.
* `n=3`, spravíme si ďalší krok. Rozoberte si všetkých 8 možností sami a mali by ste sa dostať k operáciám 
```
Z0 <- Z0 + H0,
H0 <- Z0,
H1 <- H0,
Z1 <- H1,
H2 <- H1
```
z čoho dostaneme stav `Z0 = 3, H0 = 2, H1 = 1, Z1 = 1, H2 = 1`. Tento stav zodpovedá pravdepodobnostnému rozloženiu $\tfrac{5}{8}$ pre 0 dvojíc, $\tfrac{2}{8}$ pre 1 dvojicu a $\tfrac{1}{8}$ pre 3 dvojice.
* `n=4`, spravíme si spoločne ešte jeden posledný krok. Skontrolujte si, že vám operácie a pravdepodobnosti vyšli takto:
```
Z0 <- Z0 + H0 = 5 
H0 <- Z0      = 3,
H1 <- H0 + Z1 = 3,
Z1 <- H1 + Z1 = 2,
H2 <- H1      = 1,
Z2 <- H2      = 1
H3 <- H2      = 1
====================
0 -> 8 / 16,
1 -> 5 / 16,
2 -> 2 / 16,
3 -> 1 / 16.
```

Tento postup nám síce nedáva jasný matematický výsledok, ale dáva nám recept, ako si vypočítať rozloženie pravdepodobností pre `n+1` na základe predchádzajúceho. Ak sa lepšie zapozeráme, uvidíme tam nasledovný vzorec:

$$
H_{k}(n+1) = H_{k-1}(n) + Z_{k}(n) \\
Z_{k}(n+1) = H_{k}(n) + Z_{k}(n),
$$
kde $n$ je dĺžka sekvencie a $k$ je počet dvojíc `HH`. Aj keď nevieme vzorcom vyjadriť, ako bude vyzerať ľubovoľný člen tejto postupnosti, môžme si na to napísať jednoduchý program, ktorý to zráta za nás.

In [0]:
def distribution_of_HH(N):
    if N <= 0:
        return {}
    if N == 1: # Pre malé N vrátime presný výsledok
        return {0: {'H': 1, 'Z': 1}} # stav v dĺžke 1
    
    # predchádzajúce výsledky, s ktorými urobíme indukčný krok
    # volanie samotnej funkcie z tela funkcie sa nazýva rekurzia
    # musíme si dať pozor, aby nám nevznikol nekonečný cyklus, o čo sme sa v našom prípade postarali
    previous_results = distribution_of_HH(N-1) 
    
    # pomocná funkcia na čítanie predchádzajúcich výsledkov
    def get_result(k, coin): # priklad štruktúry results je vyššie v časti N == 1.
        if k in previous_results:
            if coin in previous_results[k]:
                return previous_results[k][coin]
        return 0

    new_results = {} # sem budeme ukladať priebežné výsledky
    max_k = max(previous_results.keys()) + 1 # uvažujeme o jeden väčší počet dvojíc ako minulé max
    for k in range(max_k + 1): # po maximum vrátane
        resultH = get_result(k-1, 'H') + get_result(k, 'Z')
        resultZ = get_result(k, 'H') + get_result(k, 'Z')
        if resultH > 0 or resultZ > 0:
            new_results[k] = {'H': resultH, 'Z': resultZ}
    return new_results


In [13]:
distribution_of_HH(4) # kontrola, či to vyzerá správne

{0: {'H': 3, 'Z': 5},
 1: {'H': 3, 'Z': 2},
 2: {'H': 1, 'Z': 1},
 3: {'H': 1, 'Z': 0}}

In [14]:
def probabilities_of_HH(N):
    dist = distribution_of_HH(N)
    normalizer = 2 ** N
    return [(dist[key]['H'] + dist[key]['Z']) / normalizer for key in range(len(dist))]

probabilities_of_HH(4)

[0.5, 0.3125, 0.125, 0.0625]

In [15]:
probabilities_of_HH(100)

[7.31568061440679e-10,
 1.2440787567561787e-08,
 1.0871163664887751e-07,
 6.496777263328833e-07,
 2.9824543795551954e-06,
 1.1202431563893268e-05,
 3.5816487425611236e-05,
 0.00010014114852082952,
 0.0002496847946100435,
 0.0005634237664942142,
 0.0011639757214277253,
 0.002221872386319664,
 0.0039484453087702456,
 0.006573173413649381,
 0.010304975930365733,
 0.015282153491776654,
 0.021520853410243592,
 0.02887506660201574,
 0.037020706756839224,
 0.045471888195761596,
 0.05363013931507898,
 0.06085909899178746,
 0.06657071847879878,
 0.07030606187768916,
 0.07179520897514809,
 0.07098590284427614,
 0.06803779871937247,
 0.06328639714244333,
 0.057186185180741964,
 0.050245115231050956,
 0.042962187170072795,
 0.03577720633703758,
 0.02903784511441904,
 0.022985073156075612,
 0.01775472629226951,
 0.01339093852782089,
 0.009866437132652759,
 0.007105061281922009,
 0.00500291427035967,
 0.0034458836093027435,
 0.0023225280255036855,
 0.001532327144717912,
 0.0009899362001142612,
 0.00

Na prvý pohľad vyzerajú naše výsledky sľubne. Existuje nejaký spôsob, ako by sme si ich mohli overiť, či sú správne?

#### Overovanie výsledkov



Na štatistike je krásne, že aj keby sme nevedeli presne spočítať výsledok predchádzajúcej úlohy, môžme si ho veľmi jednoducho nasimulovať. Konkrétne si môžme začať vytvárať náhodné postupnosti dĺžky `n`, počítať na nich našu štatistiku a vykresliť si rozdelenie, ktoré sme z takéhoto experimentu dostali. Ak spravíme dostatočne veľa pokusov, mali by sme dostať riešenie veľmi podobné tomu, čo sme si spočítali. Poďme to hneď vyskúšať:

In [0]:
def count_HH(seq): ## spočítame počet dvojíc HH, teda v našich dátach 1,1
    return sum(1 if seq[i] == 1 and seq[i+1] == 1 else 0 for i in range(len(seq)-1))

def random_seq(n): ## vytvoríme náhodnú sekvenciu
    return np.random.randint(0, 2, n)

def random_HH_stat(n, samples=1000): ## spočítame našu štatistiku na náhodných dátach
    return [count_HH(random_seq(n)) for i in range(samples)]

In [17]:
N = 100

trace_histogram = go.Histogram(
    x=random_HH_stat(N, 10000),
    histnorm='probability',
    name='random histogram') 

probs = probabilities_of_HH(N)
trace_true_probs = go.Scatter(
    x=list(range(len(probs))),
    y=probs,
    mode='lines',
    name='true probability'
)
fig = go.Figure(data=[trace_histogram, trace_true_probs])
fig.update_layout(
    title="Rozdelenie počtu dvojíc HH",
    xaxis_title="Počet dvojíc HH",
    yaxis_title="Pravdepodobnosť",
)
fig.show()

Vyzerá, že sme sa celkom trafili. Do budúcnosti teda nemusíme prechádzať ťažkým matematickým cvičením rátania zložitej pravdepodobnosti, ale za pár riadkov programu si môžme celkom presne simulovať výsledok. Takáto simulácia sa odborne nazýva *Monte-carlo simulácia*, ak by ste chceli zamachrovať pred kamarátmi.

Prístup, akým sme spočítali tieto skutočné pravdepodobnosti, sa nazýva *dynamické programovanie*. Je to oblasť, kde sa pekne mixuje matematika a programovanie. Museli sme sa chvíľu matematicky zamyslieť, aké vzťahy platia v našom matematickom probléme, aby sme prišli k rekurzívnemu / indukčnému pravidlu. Toto pravidlo sme následne priamočiaro naprogramovali a dostali sme sa rýchlo k výsledku, ktorý by sme ručne počítali veľmi dlho.

Aj dynamické programovanie aj monte-carlo simulácie sú oba skvelým príkladom toho, že programovanie je skvelým a potrebným nástrojom matematikov.

#### Využitie štatistiky



Pridajme do nášho grafu histogram dvojíc `HH` vytvorený z našich dát:

In [18]:
N = 100

trace_histogram = go.Histogram(
    x=random_HH_stat(N, 10000),
    histnorm='probability',
    name='random histogram') 

probs = probabilities_of_HH(N)
trace_true_probs = go.Scatter(
    x=list(range(len(probs))),
    y=probs,
    mode='lines',
    name='true probability'
)

trace_data = go.Histogram(
    x=data.apply(count_HH),
    histnorm="probability",
    name="data histogram"
)
fig = go.Figure(data=[trace_histogram, trace_true_probs, trace_data])
fig.update_layout(
    title="Rozdelenie počtu dvojíc HH",
    xaxis_title="Počet dvojíc HH",
    yaxis_title="Pravdepodobnosť",
)
fig.show()

Všimnime si malého stĺpčeka v ľavej časti dát. Niekomu sa podarilo spraviť iba 2 postupnosti `HH` v celej sekvencii 100 hodov. Naša teória hovorí, že toto je vysoko nepravdepodobné a preto by sme mali považovať túto postupnosť za ľudsky vytvorenú.

V skutočnej štatistike sa používajú veľmi podobné argumenty, ale z opačnej strany. Skúsime nájsť takú hodnotu počtu dvojíc $x$, že šanca toho, že bude počet dvojíc v postupnosti menej ako `x` bude menej ako $0.1\%$. Rovnako skúsime nájsť takú hodnotu $y$, aby šanca toho, že bude počet dvojíc v postupnosti *viac* ako `x` bude menej ako $0.1\%$. Obe tieto hodnoty si vieme ľahko spočítať z nášho pravdepodobnostného rozdelenia. Existuje na to dokonca priamo funkcia, ktorá sa nazýva *kvantil*. S kvantilom už sme sa stretli, konkrétne $50\%$ kvantil je medián.

In [19]:
prob = 0.001
random_HH = random_HH_stat(N, 10000)
print(f'x: {np.quantile(random_HH, prob)}, y: {np.quantile(random_HH, 1 - prob)}')

x: 9.0, y: 44.0


Vyšlo nám, že $x = 9$ a $y = 44$. Pod a nad týmto číslom je menej ako 0.1\% šanca, že to nastane náhodnou sekvenciou. Aj keď je šanca malá, stále sa to mohlo náhodne stať. Preto sekvencie, ktoré sú za týmito hranicami, budeme považovať za *silno podozrivé* z toho, že sú človekom vytvorené. Napíšme si niekde tieto sekvencie, aby sme na ne nezabudli.



In [28]:
counts = data.apply(count_HH)
counts[np.logical_or(counts < 9, counts > 44)]

240     2
443    45
dtype: int64

Ide teda o sekvencie s číslami 240 a 443.

Pravdepodobnosť $0.1\%$ nie je špeciálne dôležité, konkrétne číslo sme si pravdaže vymysleli. Čím má ale sekvencia menšiu šancu na existenciu pri náhodnej voľbe, tým je *podozrivejšia*, že je ľudská. Jedna z veľmi úspešných stratégií v detekcii anomálií je spraviť veľa takýchto jednoduchých kontról, či je daný dátový bod dostatočne pravdepodobný. Ak sa vo veľa z nezávislých kontról ukáže ako nepravdepodobný, pravdepodobne ide o anomáliu, na ktorú treba upozorniť. Niečo podobné skúsime spraviť aj v ďalšej časti. Najprv si to ale skúste spraviť samostatne.

### Pokročilá štatistika


Radi by sme zopakovali náš postup z minulej časti pre viacero rôznych štatistík. Aby sme to ale nemuseli robiť celé ručne, napíšeme si kus kódu, ktorý nám to výrazne zjednoduší.

In [0]:
import plotly.figure_factory as ff
import plotly.graph_objects as go

class CoinData:
    def __init__(self):
        self.data = get_data()
    
    def compute_statistic_on_data(self, statistic):
        return self.data.apply(statistic)

    def compute_statistic_on_random(self, statistic, N=10000):
        return [statistic(np.random.randint(0, 2, len(self.data))) for i in range(N)]

    def compare_statistic(self, statistic, prob, line_height=0.1, N=10000):
        data_stats = self.compute_statistic_on_data(statistic)
        random_stats = self.compute_statistic_on_random(statistic, N)
        hist_data = go.Histogram(x=data_stats, name='data', histnorm='probability')
        hist_random = go.Histogram(x=random_stats, name='random', histnorm='probability')
        fig = go.Figure([hist_data, hist_random])

        # significances
        low, high = CoinData.get_significances(random_stats, prob)
        fig.add_traces(CoinData.traces_vertical_significances(low, high, prob, line_height))
        fig.update_layout(
            title="Rozdelenie štatistiky",
            xaxis_title="Štatistika",
            yaxis_title="Pravdepodobnosť",
        )
        data_outside = data_stats[np.logical_or(data_stats < low, data_stats > high)]
        print(f'Sekvencie mimo (1 - 2 x {prob}) intervalu:\n{data_outside}')
        return fig

    def get_outliers(self, statistic, prob):
        random_stats = self.compute_statistic_on_random(statistic, N)
        low, high = CoinData.get_significances(random_stats, prob)
        data_stats = self.compute_statistic_on_data(statistic)
        return data_stats[np.logical_or(data_stats < low, data_stats > high)].index

    @staticmethod
    def get_significances(data, prob):
        return np.quantile(data, prob), np.quantile(data, 1-prob)

    @staticmethod
    def traces_vertical_significances(low, high, prob, line_height):
        return [go.Scatter(x=[low, low], y=[0, line_height], mode='lines', name=f'low {prob}'),
                go.Scatter(x=[high, high], y=[0, line_height], mode='lines', name=f'high {prob}')]

S týmto pomocným kódom sa nám celé snaženie v minulej sekcii zredukuje na tento kus kódu.

In [22]:
cd = CoinData()
cd.compare_statistic(count_HH, 0.001)

Sekvencie mimo (1 - 2 x 0.001) intervalu:
240     2
241     9
298     9
443    45
dtype: int64


Skúsme si teraz spraviť štatistiku počtu sekvencií `HZ`:

In [23]:
def count_HZ(col):
    return sum(1 if col[i] == 1 and col[i+1] == 0 else 0 for i in range(len(col) - 1))

cd.compare_statistic(count_HZ, 0.001)

Sekvencie mimo (1 - 2 x 0.001) intervalu:
13     35
26     34
39     34
71     36
106    35
110    16
122    34
127    35
140    40
144    35
190    15
199    34
203    36
211    34
225    34
240    34
252    16
298    34
357    34
372    34
436    36
450    13
497    16
dtype: int64


Konečne sa nám podarilo identifikovať zaujímavý výsledok. Kým počet dvojíc `HH` ľudia zvládali spĺňať celkom dobre, s dvojicou `HZ` im to išlo horšie. Z grafu vieme vyčítať, že ľudia omnoho častejšie striedali hlavu na znak, ako by sa na náhodu patrilo.

Pre úplnosť sa pozrime ešte na striedanie opačným smerom.

In [24]:
def count_ZH(col):
    return sum(1 if col[i] == 0 and col[i+1] == 1 else 0 for i in range(len(col) - 1))

cd.compare_statistic(count_HZ, 0.001)

Sekvencie mimo (1 - 2 x 0.001) intervalu:
13     35
26     34
39     34
71     36
106    35
110    16
122    34
127    35
140    40
144    35
190    15
199    34
203    36
211    34
225    34
240    34
252    16
298    34
357    34
372    34
436    36
450    13
497    16
dtype: int64


Výsledky oboch týchto štatistík sú veľmi podobné. Je asi prirodzené predpokladať, že na časté striedanie hlavy a znaku musí človek často striedať aj znak a hlavu. Skúsme zobrať náš posledný výsledok a spustiť na ňom funkciu `f1_score`, aby sme vedeli, ako na tom sme.

In [25]:
outliers = cd.get_outliers(count_ZH, 0.001)
f1_score(outliers)

0.35918367346938773

To je viac ako nula, ale koľko je to v skutočnosti? Rozhodne to nie je 99%, ale je to dobrý začiatok. Pokračovanie je už ale čisto na vás. Veľa šťastia!

# Domáca úloha č. 1

Táto úloha je rovnaká ako úloha č.2. Máme ale niekoľko rozdielov. Dát je menej, sekvencie sú kratšie a boli zozbierané od matematicky zdatných jedincov. Rozdeľte ich na ľudské a náhodné.



## Podrobnosti


In [26]:
get_data_homework()

0   1   2   3   4   5   6   7   8   9   ...  40  41  42  43  44  45  46  47  48  49
0    0   1   0   1   0   0   1   1   0   0  ...   0   1   0   1   0   1   1   1   0   1
1    1   1   0   0   0   1   1   1   1   0  ...   0   1   0   1   0   1   1   1   0   1
2    0   1   0   1   0   1   0   0   0   0  ...   1   0   0   0   1   0   1   1   1   1
3    0   1   1   0   1   0   1   1   1   1  ...   0   1   1   1   1   1   1   0   0   0
4    1   1   0   0   1   1   0   0   1   0  ...   1   0   1   1   0   1   0   1   1   0
5    0   1   1   1   0   0   0   0   1   1  ...   0   1   0   1   1   0   1   0   0   1
6    0   1   1   1   1   0   0   1   1   1  ...   1   1   1   0   1   0   1   0   0   1
7    0   1   0   0   1   0   1   0   0   0  ...   1   0   0   1   1   0   1   1   1   0
8    1   0   0   1   1   1   1   0   0   1  ...   1   0   0   0   0   0   0   1   1   1
9    1   1   1   1   1   0   0   1   0   1  ...   0   1   1   1   0   1   1   1   0   1
10   0   1   1   1   0   1   1   0   1   1  ...   1   0   1   1   0   0   0   1   0   0
11   1   1   0   1   0   1   0   1   0   0  ...   0   1   0   1   0   1   0   1   0   0
12   0   1   1   0   1   1   0   1   1   1  ...   0   0   1   0   1   0   0   1   1   1
13   1   1   1   1   1   1   1   0   1   0  ...   1   1   0   1   0   1   1   1   1   1
14   0   1   1   0   1   0   1   1   1   1  ...   1   1   0   0   1   1   0   1   0   1
15   0   0   0   0   0   1   0   1   1   0  ...   1   0   0   0   1   1   0   0   1   1
16   1   0   1   0   1   0   1   0   1   0  ...   1   1   1   1   1   0   1   0   0   0
17   0   0   1   0   1   1   0   1   1   0  ...   1   0   1   1   0   0   1   1   1   0
18   1   0   0   1   1   0   1   0   1   1  ...   1   1   1   0   1   1   1   0   0   0
19   1   1   1   0   1   1   1   0   0   1  ...   0   0   0   0   0   1   1   1   0   1
20   0   1   1   1   1   0   0   0   0   1  ...   1   0   1   0   1   1   1   0   1   0
21   1   0   0   1   0   0   0   0   1   1  ...   0   0   1   1   0   1   1   0   1   1
22   1   1   1   1   0   1   1   1   1   1  ...   0   1   0   1   1   0   0   1   0   0
23   1   0   0   1   0   1   0   0   0   0  ...   0   0   1   1   1   0   1   0   1   1
24   0   0   1   0   1   0   1   1   0   0  ...   1   1   1   1   0   0   0   0   0   0
25   0   0   0   1   0   1   1   0   0   1  ...   0   0   1   1   0   1   1   0   1   0
26   1   1   1   0   0   1   1   0   1   0  ...   1   1   1   0   0   0   0   0   0   0
27   0   1   1   1   1   0   1   0   1   0  ...   1   1   1   0   1   0   1   1   1   0
28   1   1   1   0   1   0   0   1   0   0  ...   0   0   1   0   0   1   1   1   1   1
29   0   1   1   0   1   1   1   1   1   1  ...   0   1   0   0   0   1   0   0   0   1

[30 rows x 50 columns]

# Úloha pre praktikov

Ako sme spomínali, detekcia anomálií je skutočná disciplína plná reálnych problémov. V tejto úlohe dostanete dáta pacientov, ktorý majú potenciálne ochorenie známe pod názvom *thyroid*. Vašou úlohou je odseparovať zdravých pacientov od pacientov trpiacich touto chorobou.

## Podrobnosti

In [27]:
get_data_practical()

0    1    2    3    4    5   ...   15       16      17     18     19     20
0     0.45  1.0  0.0  0.0  0.0  0.0  ...  0.0  0.06100  0.0060  0.023  0.087  0.026
1     0.61  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.02900  0.0150  0.061  0.096  0.064
2     0.16  0.0  1.0  0.0  0.0  0.0  ...  0.0  0.02900  0.0190  0.058  0.103  0.056
3     0.85  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.11400  0.0030  0.024  0.061  0.039
4     0.75  1.0  0.0  0.0  0.0  0.0  ...  0.0  0.04900  0.0030  0.005  0.116  0.004
...    ...  ...  ...  ...  ...  ...  ...  ...      ...     ...    ...    ...    ...
6911  0.65  0.0  1.0  0.0  0.0  0.0  ...  0.0  0.00047  0.0080  0.154  0.075  0.204
6912  0.77  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.00006  0.0206  0.125  0.107  0.117
6913  0.41  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.00130  0.0250  0.125  0.114  0.109
6914  0.64  1.0  0.0  0.0  0.0  0.0  ...  0.0  0.00078  0.0206  0.106  0.088  0.121
6915  0.46  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.00189  0.0206  0.093  0.091  0.102

[6916 rows x 21 columns]

# Úloha pre brutálnych

Nech `x` je postupnosť hodov mincou nejakej ľubovoľnej dĺžky. Napíšte funkciu, ktorá pre `x` a prirodzené číslo `n` spočíta pravdepodobnostné rozdelenie počtu výskytu postupnosti `x` v postupnosti hodov mincou o dĺžky `n`.

Túto úlohu sme riešili pre špeciálnu voľbu `x = HH` a `n = 100` v rámci riešenia druhej úlohy.